lib/CStack.cpp

```diff
@@ -282,6 +282,13 @@ bool CStack::canBeHealed() const
 		   && !hasBonusOfType(Bonus::SIEGE_WEAPON);
 }
 
+bool CStack::isOnNativeTerrain() const
+{
+	if(base)
+		return type->isItNativeTerrain(base->armyObj->battle->terrainType);

```

Comment:
There is `this->battle` for that. Also why stack without `base` (f.e. summoned) can not have native terrain?